File: hksoho/byrydens/report/orders_due_to_pay/orders_due_to_pay.py
import frappe
from frappe import _

def execute(filters=None):
    # 使用您系統真正的欄位：requested_qty
    data = frappe.db.sql("""
        SELECT 
            YEAR(po.po_shipdate) AS year,
            MONTH(po.po_shipdate) AS month_num,
            SUM((item.requested_qty - COALESCE(item.booked_qty, 0)) * item.unit_price) AS due_amount
        FROM `tabPurchase Order` po
        JOIN `tabPurchase Order Item` item ON item.parent = po.name
        WHERE po.po_shipdate IS NOT NULL
          AND item.requested_qty > COALESCE(item.booked_qty, 0)
          AND item.requested_qty > 0
          AND item.unit_price > 0
        GROUP BY YEAR(po.po_shipdate), MONTH(po.po_shipdate)
        HAVING due_amount > 0
        ORDER BY year ASC, month_num ASC
    """, as_dict=1)

    if not data:
        return [
            {"label": "訊息", "fieldname": "msg", "fieldtype": "Data", "width": 600}
        ], [{"msg": "目前沒有任何未出貨訂單（requested_qty > booked_qty）"}]

    rows = []
    total = 0

    for d in data:
        month_name = frappe.utils.getdate(f"{d.year}-{d.month_num:02d}-01").strftime("%B %Y")
        rows.append({
            "year": d.year,
            "month": month_name,
            "due_amount": d.due_amount,
            "details": {"label": "View Details"}
        })
        total += float(d.due_amount)

    # 總計列
    rows.append({
        "year": "",
        "month": "<strong style='color:#e74c3c; font-size:16px;'>Grand Total (All Time)</strong>",
        "due_amount": total,
        "details": ""
    })

    columns = [
        {"label": "Year",       "fieldname": "year",       "fieldtype": "Int",      "width": 80},
        {"label": "Month",      "fieldname": "month",      "fieldtype": "Data",     "width": 160},
        {"label": "Amount Due", "fieldname": "due_amount", "fieldtype": "Currency", "width": 180},
        {"label": "Details",    "fieldname": "details",    "fieldtype": "Button",   "width": 130}
    ]

    return columns, rows

File: hksoho/byrydens/report/orders_due_to_pay/orders_due_to_pay.js
frappe.query_reports["Orders Due to Pay"] = {
    filters: [],  // 移除頂部 filter box
    
    onload: function(report) {
        // 隱藏所有 dropdown 和 filter 相關功能
        setTimeout(function() {
            // 隱藏 filter row
            $('.dt-row-filter').remove();
            
            // 隱藏每個欄位標題的 dropdown 按鈕
            $('.dt-dropdown').remove();
            
            // 移除 resize handle（調整欄位寬度的把手）
            $('.dt-cell__resize-handle').remove();
            
            // 隱藏 dropdown container
            $('.dt-dropdown-container').remove();
        }, 200);
    },
    formatter: function(value, row, column, data, default_formatter) {
        if (column.fieldname === "details" && data && data.details) {
            const month_name = data.month.split(' ')[0];   // 取出 "March"
            return `<button class="btn btn-xs btn-primary" 
                            onclick="showDuePODetails('${data.year}', '${month_name}')">
                    View Details
                    </button>`;
        }
        return default_formatter(value, row, column, data);
    }

};

function showDuePODetails(year, month) {
    frappe.call({
        method: "hksoho.byrydens.utils.get_due_po_details",
        args: { year: year, month_name: month },
        callback: function(r) {
            if (r.message && r.message.data && r.message.data.length > 0) {
                // 計算總額（按幣別分組）
                let currency_totals = {};
                
                r.message.data.forEach(function(row) {
                    if (row.undelivered_value) {
                        let curr = row.currency || 'Unknown';
                        currency_totals[curr] = (currency_totals[curr] || 0) + row.undelivered_value;
                    }
                });
                
                // 建立 HTML 表格
                let html = `
                    <div style="overflow-x: auto;">
                        <table class="table table-bordered table-hover" style="width: 100%; margin-top: 10px;">
                            <thead style="background-color: #f5f5f5;">
                                <tr>
                                    <th>PO Number</th>
                                    <th>Partner ID</th>
                                    <th>Partner Name</th>
                                    <th>Ship Date</th>
                                    <th>Status</th>
                                    <th>Currency</th>
                                    <th style="text-align: right;">Undelivered Value</th>
                                </tr>
                            </thead>
                            <tbody>
                `;
                
                // 加入每一行資料
                r.message.data.forEach(function(row) {
                    html += `
                        <tr>
                            <td><a href="/app/purchase-order/${row.po_number}" target="_blank">${row.po_number}</a></td>
                            <td>${row.partner_id || ''}</td>
                            <td>${row.partner_name || ''}</td>
                            <td>${frappe.datetime.str_to_user(row.po_shipdate) || ''}</td>
                            <td><span class="badge badge-primary">${row.po_status || ''}</span></td>
                            <td>${row.currency || ''}</td>
                            <td style="text-align: right;">${format_currency(row.undelivered_value, row.currency)}</td>
                        </tr>
                    `;
                });
                
                // 加入合計行
                html += `
                            </tbody>
                            <tfoot style="background-color: #f0f0f0; font-weight: bold;">
                                <tr>
                                    <td colspan="6" style="text-align: right; padding-right: 15px;">Total by Currency:</td>
                                    <td style="text-align: right;">
                `;
                
                // 顯示各幣別的合計
                Object.keys(currency_totals).sort().forEach(function(curr) {
                    html += `<div>${curr}: ${format_currency(currency_totals[curr], curr)}</div>`;
                });
                
                html += `
                                    </td>
                                </tr>
                            </tfoot>
                        </table>
                    </div>
                `;
                
                // 顯示 Dialog
                let dialog = new frappe.ui.Dialog({
                    title: r.message.title || `${month} ${year} - Orders Due to Pay`,
                    size: "extra-large",
                    fields: [{
                        fieldtype: "HTML",
                        options: html
                    }],
                    primary_action_label: "Close",
                    primary_action: function() {
                        dialog.hide();
                    }
                });
                
                dialog.show();
                
            } else {
                frappe.msgprint({
                    title: "No Data",
                    message: `No due PO found in ${month} ${year}`,
                    indicator: "blue"
                });
            }
        },
        error: function(err) {
            frappe.msgprint("API Error: " + (err.message || "Check method path"));
        }
    });
}
